### PR TITLE
Adversarial loss

### PR DESCRIPTION
**Pull request by cdancette**
- Adds `AdversarialDataLoader` (that derives from `MultiTakDataLoader`, it is just to make the y_spk between 0 and 1 instead of -1 and 1

- adds an `adversarial_classifier` object in our trainer module. (maybe we can hide it under a new argument `adversarial=False/True`

- adds a trainer:  `TrainerSiameseAdversarialLoss` where we can give a weight to the default loss

### CODE REVIEW

File: abnet3/model.py [hunk 1]
```diff
@@ -169,6 +190,27 @@ def __init__(self, input_dim=None, num_hidden_layers=None, hidden_dim=None,
         self.output_path = output_path
         self.apply(self.init_weight_method)
 
+        if self.adversarial:
+            # adversarial layer
+            adversarial_classifier = [
+                nn.Linear(2*output_dim, 100),
+                nn.ReLU(),
+                nn.Linear(100, 50),
```
Comment (Rachine) [inline abnet3/model.py]: Parameters Hard written for test?

File: abnet3/trainer.py [hunk 2]
```diff
@@ -256,6 +268,90 @@ def optimize_model(self, do_training=True):
...
+            y_spk_batch = y_spk_batch.cuda()
+            y_phn_batch = y_phn_batch.cuda()
+        emb1_batch, emb2_batch = self.network(X_batch1, X_batch2)
+        train_loss_value = self.loss(emb1_batch, emb2_batch, y_phn_batch)
+
+        # adversarial loss
+        y_spk_predictions = self.network.run_adversarial_classifier(
+            emb1_batch, emb2_batch)
+        adversarial_loss = nn.BCELoss(size_average=self.loss.avg)(y_spk_predictions, y_spk_batch)
+        return {'default': train_loss_value, 'adversarial': adversarial_loss}
```
Comment (Rachine) [inline abnet3/trainer.py]: Maybe schedule training batches with batches for spk adv and batch for phn discrimination.

advice from here : https://github.com/soumith/ganhacks#4-batchnorm
